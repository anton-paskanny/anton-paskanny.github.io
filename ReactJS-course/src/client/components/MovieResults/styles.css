.results__items {
    display: grid;
    height: 100%;
    grid-template-columns: repeat(3, 1fr);
    grid-gap: 40px;
    padding-top: 40px;
}

.results__items--fetching {
    align-items: center;
}

.results__items {
    position: relative;
}

.results__items .lds-facebook {
    grid-column: 1 / -1;
    grid-row: 1 / -1;
    margin: 0 auto;
}

.results__no-items {
    grid-column: 1 / -1;
    grid-row: 1 / -1;
    font-size: 3em;
    text-align: center;
}