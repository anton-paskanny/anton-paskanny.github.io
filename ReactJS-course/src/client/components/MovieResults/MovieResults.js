import React, { PureComponent } from 'React';

import Movie from '../Movie/Movie';
import SortPanel from './SortPanel/SortPanel';
import Spinner from '../shared/Spinner/Spinner';

import { URL_BASE } from '../../utils';

import styles from './styles.css';

class MovieResults extends PureComponent {
    state = {
        sortBy: [
            {
                name: 'release date',
                fieldToSortBy: 'release_date',
                active: true
            },
            {
                name: 'rating',
                fieldToSortBy: 'vote_average',
                active: false
            }
        ]
    }

    componentDidMount() {
        this.props.fetchMovies(`${URL_BASE}`);
    }

    componentDidUpdate(prevProps) {
        /**
         * 1. Take selected movies genres
         * 2. Create url
         * 3. Invoke action
         * 4. Update field movies (they are with the same genres as selected move has)
         */
        if (this.props.selectedMovie && (!prevProps.selectedMovie || prevProps.selectedMovie.title !== this.props.selectedMovie.title)) {
            const genres = this.props.selectedMovie.genres.map(genre => genre.toLowerCase()).join(',')
            const URL = `${URL_BASE}?filter=${genres}`;
            this.props.fetchMoviesByGenres(encodeURI(URL));
        }
    }

    renderItems() {
        if (this.props.isFetching) {
            return <Spinner />
        }

        if (this.props.movies.length === 0) {
            return <p className="results__no-items">No films found</p>
        }

        if (this.props.selectedMovie) {
            if (this.props.moviesForSelectedMovie.length) {
                 /**
                 * Delete selected movie from received movies
                 */
                const movies = this.props.moviesForSelectedMovie.filter(elem => elem.id !== this.props.selectedMovie.id);

                if (!movies.length) {
                    return <p className="results__no-items">No films found with the same genre</p>;
                }

                return movies.map(movie => {
                    return <Movie key={movie.id} data={movie} />
                });
            } else {
                return <p className="results__no-items">No films found with the same genre</p>;
            }
        }

        // For search page movies will be sorted after they
        // where received from server
        const sortedArr = this.sortMovies();
        return sortedArr.map(movie => {
            return <Movie key={movie.id} data={movie} />
        })
    }

    sortMovies() {
        const fieldToSortBy = this.state.sortBy.find(elem => elem.active).fieldToSortBy;
        
        return this.props.movies.slice().sort((a, b) => {
            if (fieldToSortBy === 'release_date') {
                return new Date(b[fieldToSortBy]) - new Date(a[fieldToSortBy]);
            }

            return b[fieldToSortBy] - a[fieldToSortBy];
        });
    }

    handleSortByChange = (e) => {
        if (e.target.classList.contains('sort-panel__filter-btn--active')) {
            return;
        }

        this.setState({
            sortBy: this.state.sortBy.map(el => {
                if (el.name === e.target.textContent) {
                    return {...el, active: !el.active};
                }

                return {...el, active: false};
            })
        });
    }

    render() {
        return (
            <div className="results">
                <SortPanel 
                           sortByConfig={this.state.sortBy}
                           handleSortByChange={this.handleSortByChange}
                           movies={this.props.movies}
                           selectedMovie={this.props.selectedMovie}
                />
                <div className={`results__items ${this.props.isFetching ? 'results__items--fetching' : ''}`}>
                    {
                        this.renderItems()
                    }
                </div>
            </div>
        )
    }
}

export default MovieResults;