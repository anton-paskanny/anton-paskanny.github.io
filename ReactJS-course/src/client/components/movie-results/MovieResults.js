import React, { PureComponent } from 'React';

import Movie from '../movie/Movie';
import SortPanel from './sortPanel/SortPanel';

import { URL_BASE } from '../../utils';

import styles from './styles.css';

class MovieResults extends PureComponent {
    state = {
        movies: []
    }

    componentDidMount() {
        const urlToFetch = `${URL_BASE}/movies`;

        fetch(urlToFetch)
        .then(res => res.json())
        .then(json => {
            this.setState({
                movies: json.data
            })
        });
    }

    renderItems() {
        if (this.state.movies.length === 0) {
            return <p className="results__no-items">No films found</p>
        }

        return this.state.movies.map(movie => {
            return <Movie key={movie.id} data={movie} />
        })
    }

    render() {
        return (
            <div className="results">
                <SortPanel />
                <div className="results__items">
                    {
                        this.renderItems()
                    }
                </div>
            </div>
        )
    }
}

export default MovieResults;