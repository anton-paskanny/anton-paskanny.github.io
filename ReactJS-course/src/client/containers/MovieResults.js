import { connect } from 'react-redux';
import { fetchMovies, fetchMoviesByGenres } from '../actions/movies';

import MovieResults from '../components/MovieResults/MovieResults';

const mapStateToProps = state => ({
    movies: state.movies.data,
    moviesForSelectedMovie: state.movies.moviesForSelectedMovie,
    selectedMovie: state.selectedMovie.data,
    isFetching: state.movies.isFetching
});

const mapDispatchToProps = dispatch => ({
    fetchMovies(url) {
        dispatch(fetchMovies(url));
    },
    fetchMoviesByGenres(url) {
        dispatch(fetchMoviesByGenres(url))
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieResults);