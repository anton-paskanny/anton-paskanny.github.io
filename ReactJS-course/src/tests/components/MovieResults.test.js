import React from 'react';
import { shallow } from 'enzyme';

import MovieResults from '../../client/components/MovieResults/MovieResults';

import { movieItem, movies } from '../mocks/dataMock';

let component, props, sortBy, event;

describe('SortPanel', () => {
    beforeEach(() => {
        sortBy = [
            {
                name: 'release date',
                fieldToSortBy: 'release_date',
                active: true
            },
            {
                name: 'rating',
                fieldToSortBy: 'vote_average',
                active: false
            }
        ];
        props = { 
            fetchMovies: jest.fn(),
            movies,
            isFetching: false
        };
        event = {
            target: {
                classList: {},
                textContent: 'rating'
            }
        }
        component = shallow(<MovieResults {...props} />);
    });
    it('render component correctly', () => {
        expect(component).toMatchSnapshot();
    });
    it('should correctly update state on handleSortByChange event', () => {
        event.target.classList.contains = jest.fn().mockReturnValue(false);
        component.instance().handleSortByChange(event);
        expect(component.state().sortBy[1].active).toBe(true);
    });
    it('should not update state on handleSortByChange event when active item was selected', () => {
        event.target.classList.contains = jest.fn().mockReturnValue(true);
        component.instance().handleSortByChange(event);
        expect(component.state().sortBy[1].active).toBe(false);
    });
    it('should display Spinner when movies are fetching', () => {
        component.setProps({ isFetching: true });
        expect(component.find('.results__items--fetching').length).toBe(1);
    });
    it('should not show any movies', () => {
        component.setProps({ movies: [] });
        expect(component.find('.results__no-items').length).toBe(1);
    });
});