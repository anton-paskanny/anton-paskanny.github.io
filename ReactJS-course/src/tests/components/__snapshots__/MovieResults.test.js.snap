// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortPanel render component correctly 1`] = `
<div
  className="results"
>
  <SortPanel
    handleSortByChange={[Function]}
    movies={
      Array [
        Object {
          "budget": 0,
          "genres": Array [
            "Action",
            "Adventure",
            "Science Fiction",
          ],
          "id": 447365,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 0,
          "vote_count": 9,
        },
        Object {
          "budget": 0,
          "genres": Array [
            "Action",
            "Science Fiction",
          ],
          "id": 320288,
          "overview": "Gathered together by Professor Charles Xavier to protect a world that fears and hates them, the X-Men had fought many battles, been on adventures that spanned galaxies, grappled enemies of limitless might, but none of this could prepare them for the most shocking struggle they would ever face. One of their own members, Jean Grey, has gained power beyond all comprehension, and that power has corrupted her absolutely! Now, they must decide if the life of the woman they cherish is worth the existence of the entire universe!",
          "poster_path": "https://image.tmdb.org/t/p/w500/6qmsupE0opYPIaBGe7T5D2FBzLs.jpg",
          "release_date": "2019-02-14",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "X-Men: Dark Phoenix",
          "vote_average": 0,
          "vote_count": 0,
        },
        Object {
          "budget": 0,
          "genres": Array [
            "Comedy",
            "Drama",
            "Thriller",
          ],
          "id": 421044,
          "overview": "When a drifter befriends a quirky mortician, an unlikely business partnership is formed. Paranoia soon develops, however, and both men are forced to come to terms with the fragility of friendship and loyalty.",
          "poster_path": "https://image.tmdb.org/t/p/w500/5Knp0sBWb4hLTc49qaiDjZ4DYV8.jpg",
          "release_date": "2018-03-30",
          "revenue": 0,
          "runtime": 108,
          "tagline": "Friendship before money",
          "title": "Best F(r)iends",
          "vote_average": 3.3,
          "vote_count": 9,
        },
        Object {
          "budget": 55000000,
          "genres": Array [
            "Adventure",
            "Drama",
            "Western",
          ],
          "id": 384680,
          "overview": "A legendary Native American-hating Army captain nearing retirement in 1892 is given one last assignment: to escort a Cheyenne chief and his family through dangerous territory back to his Montana reservation.",
          "poster_path": "https://image.tmdb.org/t/p/w500/rqoezyB51GfhiloOB5ZErg5HXas.jpg",
          "release_date": "2017-12-22",
          "revenue": 29790924,
          "runtime": 133,
          "tagline": "No peace without sacrifice",
          "title": "Hostiles",
          "vote_average": 6.1,
          "vote_count": 272,
        },
      ]
    }
    sortByConfig={
      Array [
        Object {
          "active": true,
          "fieldToSortBy": "release_date",
          "name": "release date",
        },
        Object {
          "active": false,
          "fieldToSortBy": "vote_average",
          "name": "rating",
        },
      ]
    }
  />
  <div
    className="results__items "
  >
    <Movie
      data={
        Object {
          "budget": 0,
          "genres": Array [
            "Action",
            "Adventure",
            "Science Fiction",
          ],
          "id": 447365,
          "overview": "The third film based on Marvel's Guardians of the Galaxy.",
          "poster_path": "https://image.tmdb.org/t/p/w500/ldoY4fTZkGISMidNw60GHoNdgP8.jpg",
          "release_date": "2020-05-01",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "Guardians of the Galaxy Vol. 3",
          "vote_average": 0,
          "vote_count": 9,
        }
      }
      key="447365"
    />
    <Movie
      data={
        Object {
          "budget": 0,
          "genres": Array [
            "Action",
            "Science Fiction",
          ],
          "id": 320288,
          "overview": "Gathered together by Professor Charles Xavier to protect a world that fears and hates them, the X-Men had fought many battles, been on adventures that spanned galaxies, grappled enemies of limitless might, but none of this could prepare them for the most shocking struggle they would ever face. One of their own members, Jean Grey, has gained power beyond all comprehension, and that power has corrupted her absolutely! Now, they must decide if the life of the woman they cherish is worth the existence of the entire universe!",
          "poster_path": "https://image.tmdb.org/t/p/w500/6qmsupE0opYPIaBGe7T5D2FBzLs.jpg",
          "release_date": "2019-02-14",
          "revenue": 0,
          "runtime": null,
          "tagline": "",
          "title": "X-Men: Dark Phoenix",
          "vote_average": 0,
          "vote_count": 0,
        }
      }
      key="320288"
    />
    <Movie
      data={
        Object {
          "budget": 0,
          "genres": Array [
            "Comedy",
            "Drama",
            "Thriller",
          ],
          "id": 421044,
          "overview": "When a drifter befriends a quirky mortician, an unlikely business partnership is formed. Paranoia soon develops, however, and both men are forced to come to terms with the fragility of friendship and loyalty.",
          "poster_path": "https://image.tmdb.org/t/p/w500/5Knp0sBWb4hLTc49qaiDjZ4DYV8.jpg",
          "release_date": "2018-03-30",
          "revenue": 0,
          "runtime": 108,
          "tagline": "Friendship before money",
          "title": "Best F(r)iends",
          "vote_average": 3.3,
          "vote_count": 9,
        }
      }
      key="421044"
    />
    <Movie
      data={
        Object {
          "budget": 55000000,
          "genres": Array [
            "Adventure",
            "Drama",
            "Western",
          ],
          "id": 384680,
          "overview": "A legendary Native American-hating Army captain nearing retirement in 1892 is given one last assignment: to escort a Cheyenne chief and his family through dangerous territory back to his Montana reservation.",
          "poster_path": "https://image.tmdb.org/t/p/w500/rqoezyB51GfhiloOB5ZErg5HXas.jpg",
          "release_date": "2017-12-22",
          "revenue": 29790924,
          "runtime": 133,
          "tagline": "No peace without sacrifice",
          "title": "Hostiles",
          "vote_average": 6.1,
          "vote_count": 272,
        }
      }
      key="384680"
    />
  </div>
</div>
`;
